dirty_equals/pytest_plugin.py

```diff
@@ -0,0 +1,285 @@
+from __future__ import annotations as _annotations
+
+import ast
+import os
+import sys
+import textwrap
+from contextvars import ContextVar
+from dataclasses import dataclass
+from enum import Enum
+from functools import lru_cache
+from itertools import groupby
+from pathlib import Path
+from types import FrameType
+from typing import TYPE_CHECKING, Any, Callable, Generator, Sized
+
+import pytest
+from executing import Source
+
+if TYPE_CHECKING:
+    from _pytest.fixtures import SubRequest
+
+__all__ = ('insert_assert',)
+
+
+@dataclass
+class ToReplace:
+    file: Path
+    start_line: int
+    end_line: int
+    code: str
+
+
+to_replace: list[ToReplace] = []
+insert_assert_calls: ContextVar[int] = ContextVar('insert_assert_calls', default=0)
+insert_assert_print: ContextVar[bool] = ContextVar('insert_assert_print')
+insert_assert_enabled: ContextVar[bool] = ContextVar('insert_assert_enabled')
+
+
+def insert_assert(value: Any) -> int:
+    call_frame: FrameType = sys._getframe(1)
+    if sys.version_info < (3, 8):  # pragma: no cover
+        raise RuntimeError('insert_assert() requires Python 3.8+')
+    if not insert_assert_enabled.get():
+        raise RuntimeError('insert_assert() is disabled, either due to --insert-assert-disable or "CI" env var')
+
+    format_code = load_black()
+    ex = Source.for_frame(call_frame).executing(call_frame)
+    if ex.node is None:  # pragma: no cover
+        python_code = format_code(str(custom_repr(value)))
+        raise RuntimeError(
+            f'insert_assert() was unable to find the frame from which it was called, called with:\n{python_code}'
+        )
+    ast_arg = ex.node.args[0]
+    if isinstance(ast_arg, ast.Name):
+        arg = ast_arg.id
+    else:
+        arg = ' '.join(map(str.strip, ex.source.asttokens().get_text(ast_arg).splitlines()))
+
+    python_code = format_code(f'# insert_assert({arg})\nassert {arg} == {custom_repr(value)}')
+
+    python_code = textwrap.indent(python_code, ex.node.col_offset * ' ')
+    to_replace.append(ToReplace(Path(call_frame.f_code.co_filename), ex.node.lineno, ex.node.end_lineno, python_code))
+    calls = insert_assert_calls.get() + 1
+    insert_assert_calls.set(calls)
+    return calls
+
+
+def pytest_addoption(parser: Any) -> None:
+    parser.addoption(
+        '--insert-assert-print',
+        action='store_true',
+        default=False,
+        help='Print statements that would be substituted for insert_assert(), instead of writing to files',
+    )
+    parser.addoption(
+        '--insert-assert-fail',
+        action='store_true',
+        default=False,
+        help='Fail tests which include one or more insert_assert() calls',
+    )
```

Comment:
I tried that, for me at least it was very annoying.

I guess at least we should have an env var that means you can change the behavior for an entire session.

Reply:
Maybe a config in pyproject.toml:
```toml
[tool.insertassert]
pass: true
````

I do see the value in having it pass by default. But I feel strongly that failing if we change any code should be the default behavior.

> I tried that, for me at least it was very annoying.

What was the annoying part? I think you'd run it once get a bunch of noisy failures, look at the diff and be like "alright this is okay" -> accept the changes.

I like how insta takes it one step further and has a `generate -> review -> apply` stage. Maybe we could do something like:
```python
# insert_assert(resp.json())
assert resp.json() == {"foo": "bar"}
```
Run with `--insert-assert-replace`
```python
# insert_assert(resp.json())
# proposed: assert resp.json() == {}
assert resp.json() == {"foo": "bar"}
```
That test run fails and since the original assertion was not replaced subsequent test runs fail by default unless you run with`--insert-assert-preview` which uses the `proposed` version.
Once you've reviewed the proposed changes (maybe via `git diff`, although we could come up with some fancy tools) you run `--insert-assert-accept` which yields:
```python
# insert_assert(resp.json())
assert resp.json() == {}
```
Alternatively you could just run with `--insert-assert-accept` and `--insert-assert-pass` from the starting point and just get to the end result with no failing tests directly.

Reply:
On `[tool.insertassert]`, we can just use [pytest's `addopts`](https://docs.pytest.org/en/6.2.x/customize.html#pytest-ini). But with that pytest will fail if dirty-equals (or devtools, see below) is not installed.

I'm happy to add more features over time, but for initial release I'd prefer to keep it as simple as possible.

I'd be happy to add `--insert-assert-pass` as well as `--insert-assert-fail` and default to `--insert-assert-fail`, that way I'll just set `--insert-assert-pass` myself.

The option option is to add a new result status, see [`pytest_report_teststatus`](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.hookspec.pytest_report_teststatus), problem is I tried that and pytest-sugar didn't know how to deal with the custom status.

Then again, pytest-sugar needs some love I'm worried I might have to stop using it.

Reply:
on pytest-sugar https://github.com/Teemu/pytest-sugar/issues/224